packages/apollo-server-core/src/plugin/usageReporting/plugin.ts

```diff
@@ -608,36 +627,57 @@ export function ApolloServerPluginUsageReporting<TContext>(
               }
             }
 
-            function getTraceSignature(): string {
+            // Calculates signature and referenced fields for the current document.
+            // Only call this when the document properly parses and validates and
+            // the given operation name (if any) is known!
+            function getOperationDerivedData(): OperationDerivedData {
               if (!requestContext.document) {
                 // This shouldn't happen: no document means parse failure, which
                 // uses its own special statsReportKey.
                 throw new Error('No document?');
               }
 
-              const cacheKey = signatureCacheKey(
+              const cacheKey = operationDerivedDataCacheKey(
                 requestContext.queryHash,
                 requestContext.operationName || '',
               );
 
+              // Ensure that the cache we have is for the right schema.
+              if (
+                !operationDerivedDataCache ||
+                operationDerivedDataCache.forSchema !== schema
+              ) {
+                operationDerivedDataCache = {
+                  forSchema: schema,
+                  cache: createOperationDerivedDataCache({ logger }),
+                };
+              }
+
               // If we didn't have the signature in the cache, we'll resort to
               // calculating it.
-              const cachedSignature = signatureCache.get(cacheKey);
-
-              if (cachedSignature) {
-                return cachedSignature;
+              const cached = operationDerivedDataCache.cache.get(cacheKey);
+              if (cached) {
+                return cached;
               }
 
               const generatedSignature = (
                 options.calculateSignature || defaultUsageReportingSignature
               )(requestContext.document, requestContext.operationName || '');
 
+              const generated: OperationDerivedData = {
```

Comment:
Maybe just name it `operationDerivedData`? `generated` a bit ambiguous

Reply:
The theory was that this is "generated (now)" vs "cached" but sure.

Reply:
f483e79